catalog/migrate/workspace-query-separation/src/main/java/ddf/catalog/migrate/migration/workspace/query/separation/WorkspaceQueryMigration.java

```diff
@@ -0,0 +1,214 @@
+            .map(String.class::cast)
+            .map(xml -> xml.getBytes(StandardCharsets.UTF_8))
+            .map(ByteArrayInputStream::new)
+            .map(this::xmlToMetacard)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+
+    LOGGER.trace("Creating query metacards for workspace [{}]", workspaceMetacard.getId());
+
+    CreateResponse response = createMetacards(queryMetacards);
+
+    if (response != null) {
+      List<Serializable> createdMetacardIds =
+          response
+              .getCreatedMetacards()
+              .stream()
+              .map(Metacard::getId)
+              .filter(Objects::nonNull)
+              .collect(Collectors.toList());
+
+      LOGGER.trace("Created query metacards with IDs {}", createdMetacardIds);
+
+      workspaceMetacard.setAttribute(new AttributeImpl(WORKSPACE_QUERIES, createdMetacardIds));
+
+      updateMetacard(workspaceMetacard);
+
+      LOGGER.trace(
+          "Updated workspace metacard [{}] with query metacard associations {}",
+          workspaceMetacard.getId(),
+          createdMetacardIds);
+    } else {
+      LOGGER.debug("An error occurred while creating query metacards");
+    }
+  }
+
+  private UpdateResponse updateMetacard(Metacard metacard) {
+    return AccessController.doPrivileged(
+        (PrivilegedAction<UpdateResponse>)
+            () -> {
+              try {
+                return catalogFramework.update(new UpdateRequestImpl(metacard.getId(), metacard));
+              } catch (IngestException | SourceUnavailableException e) {
+                LOGGER.debug("Error updating workspace metacard [{}]", metacard.getId(), e);
+              }
+              return null;
+            });
+  }
+
+  private CreateResponse createMetacards(List<Metacard> metacards) {
+    return AccessController.doPrivileged(
+        (PrivilegedAction<CreateResponse>)
+            () -> {
+              try {
+                return catalogFramework.create(new CreateRequestImpl(metacards));
+              } catch (IngestException | SourceUnavailableException e) {
+                LOGGER.debug("Error while creating query metacards", e);
+              }
+              return null;
+            });
+  }
+
+  private Metacard xmlToMetacard(InputStream inputStream) {
+    try {
+      Metacard metacard = xmlInputTransformer.transform(inputStream);
+      metacard.setAttribute(new AttributeImpl(Core.METACARD_TAGS, QUERY_TAG));
+
+      LOGGER.trace("Successfully parsed query metacard with id [{}]", metacard.getId());
+
+      return metacard;
+    } catch (IOException | CatalogTransformerException e) {
+      LOGGER.debug("Error parsing query metacard input stream", e);
+
+      return null;
+    } finally {
+      try {
+        inputStream.close();
+      } catch (IOException e) {
+        LOGGER.debug("An error occurred while closing the input stream", e);
+      }
+    }
+  }
+
+  @VisibleForTesting
+  void setSecurity(Security security) {
```

Comment:
✏️ This could be injected via a constructor parameter instead. You'd just have to create a `Security` bean using `factory-method="getInstance"` in your blueprint.